File: backend/app/api/routes/projects.py
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_user, get_db
from app.db.models.user import User
from app.schemas.project import (
    ProjectCreate,
    ProjectListResponse,
    ProjectRead,
    ProjectUpdate,
)
from app.services.project_service import project_service

router = APIRouter(prefix="/projects", tags=["Projects"])


@router.post(
    "",
    response_model=ProjectRead,
    status_code=status.HTTP_201_CREATED,
)
def create_project(
    payload: ProjectCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return project_service.create_project(db, payload, current_user)


@router.get("", response_model=ProjectListResponse)
def list_projects(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    projects = project_service.list_projects(db, current_user)

    return ProjectListResponse(
        success=True,
        count=len(projects),
        projects=projects,
    )


@router.get("/{project_id}", response_model=ProjectRead)
def get_project(
    project_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    project = project_service.get_project_by_id(db, project_id, current_user)

    if not project:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Project not found",
        )

    return project


@router.patch("/{project_id}", response_model=ProjectRead)
def update_project(
    project_id: UUID,
    payload: ProjectUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    project = project_service.get_project_by_id(db, project_id, current_user)

    if not project:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Project not found",
        )

    return project_service.update_project(db, project, payload)


@router.delete("/{project_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_project(
    project_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    project = project_service.get_project_by_id(db, project_id, current_user)

    if not project:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Project not found",
        )

    project_service.soft_delete_project(db, project)

    return None

File: frontend/src/api/projects.ts
import { apiClient } from "./client";

export type CreateProjectPayload = {
  name: string;
  description?: string;
};

export async function listProjects() {
  const response = await apiClient.get("/projects");
  return response.data;
}

export async function createProject(payload: CreateProjectPayload) {
  const response = await apiClient.post("/projects", payload);
  return response.data;
}

export async function getProjectDashboardSummary(projectId: string) {
  const response = await apiClient.get(`/projects/${projectId}/dashboard/summary`);
  return response.data;
}

export type UpdateProjectPayload = {
  name?: string;
  description?: string;
};

export async function updateProject(projectId: string, payload: UpdateProjectPayload) {
  const response = await apiClient.patch(`/projects/${projectId}`, payload);
  return response.data;
}

export async function deleteProject(projectId: string) {
  await apiClient.delete(`/projects/${projectId}`);
}